gridnet/gridnet.py

```diff
@@ -0,0 +1,168 @@
+        self.downsample_specs = []
+        for i in range(self.height):
+            row_lateral_specs, row_upsample_specs, row_downsample_specs = [], [], []
+            num_filters = channel_sizes[i]
+            common_spec = [num_filters, 1]
+            row_lateral_specs = [common_spec for j in range(self.num_lateral_convs)]
+
+            if i > 0:
+                row_downsample_specs = [common_spec for j in range(self.num_downsampling_convs)]
+
+            if i < self.height - 1:
+                row_upsample_specs = [common_spec for j in range(self.num_upsample_convs)]
+
+            self.lateral_specs.append(row_lateral_specs)
+            self.upsample_specs.append(row_upsample_specs)
+            self.downsample_specs.append(row_downsample_specs)
+
+    def get_forward(self, features, reuse_variables=False):
+        """
+        :param features: A Tensor. Input feature maps of shape [batch_size, H, W, num_features]
+        :return: final_output: A Tensor. Will take on the same shape as param features.
+                 grid_outputs: A 2D list of Tensors. Represents the output at each grid node.
+        """
+
+        with tf.variable_scope(self.name, reuse=reuse_variables):
+            grid_outputs = [[None for x in range(self.width)] for y in range(self.height)]
+
+            # First lateral connection.
+            grid_outputs[0][0] = self.process_rightwards(features, 0, 0)
+
+            # Connect first half (Down-sampling streams) by iterating to the right, and downwards.
+            for i in range(self.height):
+                for j in range(int(self.width / 2)):
+                    if i == 0 and j == 0:
+                        continue
+
+                    top_output, left_output = 0, 0
+                    if i > 0:
+                        top_output = self.process_downwards(grid_outputs[i-1][j], i, j)
+                    if j > 0:
+                        left_output = self.process_rightwards(grid_outputs[i][j-1], i, j)
+
+                    grid_outputs[i][j] = top_output + left_output
+
+            # Connect second half (Up-sampling streams) by iterating to the right, and upwards.
+            for i in range(self.height - 1, -1, -1):
+                for j in range(int(self.width / 2), self.width):
+                    bottom_output, left_output = 0, 0
+                    if i < self.height - 1:
+                        bottom_output = self.process_upwards(grid_outputs[i+1][j], i, j)
+                    if j > 0:
+                        left_output = self.process_rightwards(grid_outputs[i][j-1], i, j)
+
+                    grid_outputs[i][j] = bottom_output + left_output
+
+            # Final lateral connection.
+            previous_output = grid_outputs[0][self.width-1]
+            final_output = self.process_rightwards(previous_output, 0, self.width)
+            return final_output, grid_outputs
+
+    # Helper functions.
+    def process_rightwards(self, input, i, j):
```

Comment:
For private functions, use _process_rightwards with an extra _ at the front.

Reply:
Will do.